include/quicly/issued_cid.h

```diff
@@ -0,0 +1,155 @@
+ *
+ * @return 0 if successfully generated, non-zero if unable to generate a CID (e.g. limit reached)
+ */
+typedef int (*quicly_issued_cid_generator_t)(quicly_conn_t *conn, quicly_issued_cid_t *cid);
+
+enum en_quicly_issued_cid_state_t {
+    /**
+     * this entry is free for use
+     */
+    QUICLY_ISSUED_CID_STATE_IDLE,
+    /**
+     * this entry is to be sent at the next round of send operation
+     */
+    QUICLY_ISSUED_CID_STATE_PENDING,
+    /**
+     * this entry has been sent and is waiting for ACK (or to be deemed lost)
+     */
+    QUICLY_ISSUED_CID_STATE_INFLIGHT,
+    /**
+     * this CID has been delivered to the peer (ACKed) and in use
+     */
+    QUICLY_ISSUED_CID_STATE_DELIVERED,
+};
+
+/**
+ * records information for sending NEW_CONNECTION_ID frame
+ */
+struct st_quicly_issued_cid_t {
+    enum en_quicly_issued_cid_state_t state;
+    uint64_t sequence;
+    quicly_cid_t cid;
+    uint8_t stateless_reset_token[QUICLY_STATELESS_RESET_TOKEN_LEN];
+};
+
+/**
+ * manages a list of connection IDs we issue to the peer
+ */
+struct st_quicly_issued_cid_set_t {
+    /**
+     * storage to retain issued CIDs
+     *
+     * Pending CIDs (state == STATE_PENDING) are moved to the front of the array, in the order it was marked as pending.
+     * This ensures that pending CIDs are sent in FIFO manner. Order of CIDs with other states is not defined.
+     */
+    struct st_quicly_issued_cid_t cids[QUICLY_LOCAL_ACTIVE_CONNECTION_ID_LIMIT];
+    /**
+     * how many entries are actually usable in `cids`?
+     */
+    size_t _capacity;
+    quicly_issued_cid_generator_t _generator;
+    /**
+     * connection object passed to the CID generator
+     */
+    quicly_conn_t *_conn;
+};
+
+/**
+ * initialize the structure
+ *
+ * If `generator` is non-NULL, it is initialized with capacity==1 (sequence==0 is registered as DELIVERED).
+ * Otherwise, it is initialized with capacity==0, and the capacity shall never be increased.
+ */
+void quicly_issued_cid_init(quicly_issued_cid_set_t *set, quicly_issued_cid_generator_t generator, quicly_conn_t *conn);
+/**
+ * sets a new capacity of issued CIDs.
+ *
+ * The new capacity must be equal to or grater than the current capacity, and must be equal to or less than the elements of `cids`.
+ * When the capacity is expanded, the CID generator callback is called to generate a new CID.
+ */
+void quicly_issued_cid_set_capacity(quicly_issued_cid_set_t *set, size_t new_cap);
+/**
+ * returns true if all entries in the given set is in IDLE state
+ */
+static int quicly_issued_cid_is_empty(const quicly_issued_cid_set_t *set);
+static size_t quicly_issued_cid_get_capacity(const quicly_issued_cid_set_t *set);
+/**
+ * mark the first `num_sent` pending CIDs as INFLIGHT.
+ */
+void quicly_issued_cid_mark_inflight(quicly_issued_cid_set_t *set, size_t num_sent);
+/**
+ * mark the specified CID as DELIVERED.
+ *
+ * @return zero if successful, non-zero if the specified CID was not found.
```

Comment:
Our habit is to use 1 for success or 0 for failure when the function returns a boolean. Though I think this function can be `void`?

Reply:
I agree that this can be of type `void`.

Just for future reference, my understanding of the convention so far is that:
* returns boolean => 1 for success, 0 for failure
* returns error code => 0 for success, non-zero (mostly positive?) for failure

Reply:
After refactoring, this function has become a boolean, with true indicating "there is something to send".